jsonrpcservice/__init__.py

```diff
@@ -0,0 +1,259 @@
+import abc
+import inspect
+import json
+
+
+class JsonRpcError(Exception):
+    def __init__(self, code, message, data=None):
+        super().__init__(message)
+        self.code = code
+        self.message = message
+        self.data = data
+
+
+class ParseError(JsonRpcError):
+    code = -32700
+
+    def __init__(self, message="Parse error", data=None):
+        super().__init__(self.code, message, data)
+
+
+class InvalidRequest(JsonRpcError):
+    code = -32600
+
+    def __init__(self, message="Invalid request", data=None):
+        super().__init__(self.code, message, data)
+
+
+class MethodNotFound(JsonRpcError):
+    code = -32601
+
+    def __init__(self, message="Method not found", data=None):
+        super().__init__(self.code, message, data)
+
+
+class InvalidParams(JsonRpcError):
+    code = -32602
+
+    def __init__(self, message="Invalid params", data=None):
+        super().__init__(self.code, message, data)
+
+
+class InternalError(JsonRpcError):
+    code = -32603
+
+    def __init__(self, message="Internal error", data=None):
+        super().__init__(self.code, message, data)
+
+
+class ApplicationError(JsonRpcError):
+    def __init__(self, code, message="Application error", data=None):
+        if -32000 >= code >= -32768:
+            raise ValueError("Error codes between -32000 and -32768 are reserved by JSON-RPC 2.0 specification.")
+
+        super().__init__(code, message, data)
+
+
+class Response(object, metaclass=abc.ABCMeta):
+    def __init__(self, request):
+        self.request = request
+
+    @abc.abstractmethod
+    def dict(self):
+        pass
+
+    def body(self):
+        if self.request.is_notification():
+            return ""
+        else:
+            return json.dumps(self.dict())
+
+
+class ErrorResponse(Response):
+    def __init__(self, request, code, message, data=None, exception=None):
+        super().__init__(request)
+        self.code = code
+        self.message = message
+        self.data = data
+        self.exception = exception
+
+    def dict(self):
+        error = {
+            'message': self.message,
+            'code': self.code,
+        }
+
+        if self.data is not None:
+            error['data'] = self.data
+
+        return {
+            "jsonrpc": "2.0",
+            "id": self.request.id(),
+            "error": error
+        }
+
+
+class SuccessResponse(Response):
+    def __init__(self, request, result):
+        super().__init__(request)
+        self.result = result
+
+    def dict(self):
+        return {
+            "jsonrpc": "2.0",
+            "id": self.request.id(),
+            "result": self.result
+        }
+
+
+class Request(object):
+    def __init__(self, raw_request=None, parsed_request=None):
+        self._id = None
+        self._method = None
+        self._args = []
+        self._kwargs = {}
+        self._notification = False
+        self._prepared = False
+
+        if raw_request is None and parsed_request is None:
```

Comment:
Wouldn't it be more convenient to pass only one argument containing a request and handle it as raw or parsed depending on its type?

Reply:
```
>>> json.loads('"foo"')
'foo'
```
Both is string, this might cause some unintended things.

Reply:
I thought of this and in general you are right of course. But isn't it safe to assume, that the request is always an object for JSON-RPC requests?

Reply:
You cannot assume what a client will send you. And there is a difference between ParseError and InvalidRequest (as per specification) and this shortcut would make it impossible to safely differentiate between them. I see no harm in being explicit here and don't attempt automagic.